File: src/main/java/com/arclights/musiclights/gui/LightControl.java
package com.arclights.musiclights.gui;

import com.arclights.musiclights.core.LightRig;
import javafx.geometry.HPos;
import javafx.geometry.Orientation;
import javafx.scene.control.Button;
import javafx.scene.control.Control;
import javafx.scene.control.Slider;
import javafx.scene.control.TitledPane;
import javafx.scene.layout.GridPane;

public class LightControl extends GridPane {
    private boolean powerOn = true;

    public LightControl(int controlNbr, LightRig lightRig, String shortcut) {
        setVgap(25);

        Button powerButton = createPowerButton(lightRig, controlNbr);
        Button resetButton = createResetButton();
        Control slider = createSlider(lightRig, controlNbr, resetButton);

        add(powerButton, 0, 0);
        add(resetButton, 0, 1);
        add(slider, 0, 2);
    }

    private Button createPowerButton(LightRig lightRig, int controlNbr) {
        Button power = new Button("OFF");
        power.setOnAction(event -> {
            if (powerOn) {
                powerOn = false;
                lightRig.turnPowerOff(controlNbr);
                power.setText("ON");
            } else {
                powerOn = true;
                lightRig.turnPowerOn(controlNbr);
                power.setText("OFF");
            }
        });
        GridPane.setHalignment(power, HPos.CENTER);

        return power;
    }

    private Button createResetButton() {
        Button reset = new Button("Reset");
        GridPane.setHalignment(reset, HPos.CENTER);

        return reset;
    }

    private Control createSlider(LightRig lightRig, int controlNbr, Button resetButton) {
        Slider slider = new Slider(5, 20, 10);
        slider.setMajorTickUnit(5);
        slider.setMinorTickCount(1);
        slider.setShowTickMarks(true);
        slider.setShowTickLabels(true);
        slider.setOrientation(Orientation.VERTICAL);
        slider.valueProperty().addListener((obsVal,oldVal,newVal) -> lightRig.setLevel(controlNbr, newVal.floatValue() * 0.1f));
        TitledPane sliderPane = new TitledPane("Amplification", slider);
        sliderPane.setCollapsible(false);
        resetButton.setOnAction(event -> slider.setValue(10));

        return sliderPane;
    }

}

File: src/main/kotlin/com/arclights/musiclights/gui/light/LightControl.kt
package com.arclights.musiclights.gui.light

import com.arclights.musiclights.core.Light
import javafx.event.EventHandler
import javafx.geometry.HPos
import javafx.geometry.Orientation
import javafx.geometry.Pos
import javafx.scene.control.Button
import javafx.scene.control.Slider
import javafx.scene.control.TitledPane
import javafx.scene.layout.GridPane
import javafx.scene.layout.VBox
import javafx.scene.paint.Color

class LightControl(light: Light, color: Color) : GridPane() {

    init {
        vgap = 25.0

        val freqView = FrequencyView(light, color)
        freqView.widthProperty().bind(widthProperty())
        val lightView = LightView(light, color)
        lightView.widthProperty().bind(widthProperty())
        val powerButton = PowerButton(light)
        val slider = AmplificationSlider(light)

        add(freqView, 0, 0)
        add(lightView, 0, 1)
        add(powerButton, 0, 2)
        add(slider, 0, 3)
    }
}

class PowerButton(light: Light) : Button() {
    private var powerOn = true

    init {
        text = "OFF"
        onAction = EventHandler {
            if (powerOn) {
                powerOn = false
                light.turnOff()
                text = "ON"
            } else {
                powerOn = true
                light.turnOn()
                text = "OFF"
            }
        }
        GridPane.setHalignment(this, HPos.CENTER)
    }
}

class ResetButton(slider: Slider) : Button() {
    init {
        text = "Reset"
        GridPane.setHalignment(this, HPos.CENTER)
        onAction = EventHandler { slider.value = 10.0 }
    }
}

class AmplificationSlider(light: Light) : TitledPane() {
    init {
        val slider = Slider(5.0, 20.0, 10.0)
        slider.majorTickUnit = 5.0
        slider.minorTickCount = 1
        slider.isShowTickMarks = true
        slider.isShowTickLabels = true
        slider.orientation = Orientation.VERTICAL
        slider.valueProperty()
            .addListener { _, _, newVal -> light.setIndividualAdjustmentLevel(newVal.toFloat() * 0.1f) }

        val box = VBox()
        box.children.addAll(ResetButton(slider), slider)
        box.alignment = Pos.CENTER

        text = "Amplifier"
        content = box
        isCollapsible = false
    }
}